src/colmap/estimators/cost_functions.h

```diff
@@ -47,27 +47,83 @@ template <typename T>
 using EigenQuaternionMap = Eigen::Map<const Eigen::Quaternion<T>>;
 using EigenMatrix6d = Eigen::Matrix<double, 6, 6>;
 
+enum CovarianceType {
+  IDENTITY = 1,
+  DIAGONAL = 2,
+  GENERAL = 3,
+};
+
+template <CovarianceType CTYPE = CovarianceType::GENERAL>
 inline Eigen::MatrixXd SqrtInformation(const Eigen::MatrixXd& covariance) {
-  return covariance.inverse().llt().matrixL().transpose();
+  if constexpr (CTYPE == CovarianceType::IDENTITY)
+    return covariance;
+  else
+    return covariance.inverse().llt().matrixL().transpose();
+}
+
+inline CovarianceType GetCovarianceType(const Eigen::MatrixXd& covariance) {
+  if (covariance.isDiagonal()) {
+    if (covariance.isIdentity())
+      return CovarianceType::IDENTITY;
+    else
+      return CovarianceType::DIAGONAL;
+  } else
+    return CovarianceType::GENERAL;
+}
+
+template <CovarianceType CTYPE>
+inline bool CheckCovarianceByType(const Eigen::MatrixXd& covariance) {
+  if constexpr (CTYPE == CovarianceType::IDENTITY)
+    return covariance.isIdentity();
+  else if constexpr (CTYPE == CovarianceType::DIAGONAL)
+    return covariance.isDiagonal();
+  else
+    return true;
+}
+
+template <typename T, int N, CovarianceType CTYPE = CovarianceType::IDENTITY>
+inline void ApplyCovariance(T* residuals,
+                            const Eigen::MatrixXd& sqrt_information) {
```

Comment:
```suggestion
                            const Eigen::Matrix<double, N, N>& sqrt_information) {
```
... then we also don't need the THROW_CHECK below?

Currently, I think, this will also incur a copy of the matrix, because Eigen::Matrix2d/3d lives on the stack and Eigen::MatrixXd lives on the heap.